src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,250 @@
+        {
+            unchecked
+            {
+                fixed (uint* state = _state)
+                {
+                    var queue = state + Queue;
+                    var position = state[Length] & 0x3;
+
+                    // Mix the accumulators (if they contain anything)
+
+                    var hash = state[Length] > 3
+                        ? Rol(state[0], 1) + Rol(state[1], 7) + Rol(state[2], 12) + Rol(state[3], 18)
+                        : Prime5;
+
+                    hash += state[Length] * 4;
+
+                    // Mix what remains in the queue
+
+                    for (var i = 0; i < position; i++)
+                    {
+                        hash += queue[i] * Prime3;
+                        hash = Rol(hash, 17) * Prime4;
+                    }
+
+                    // Final mix
+
+                    hash ^= hash >> 15;
+                    hash *= Prime2;
+                    hash ^= hash >> 13;
+                    hash *= Prime3;
+                    hash ^= hash >> 16;
+
+                    return (int)hash;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        private uint Round(uint seed, uint input)
+        {
+            unchecked
+            {
+                seed += input * Prime2;
+                seed = Rol(seed, 13);
+                seed *= Prime1;
+                return seed;
+            }
+        }
+
+#       pragma warning disable 0809
+        [Obsolete("Use ToHashCode to retrieve the computed hash code.", error: true)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode() => ToHashCode();
+#       pragma warning restore 0809
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        private static uint Rol(uint value, int count)
+            => (value << count) | (value >> (32 - count));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        public void Add<T>(T value) => Add(value?.GetHashCode() ?? 0);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        public void Add<T>(T value, IEqualityComparer<T> comparer)
+        {
+            if (comparer is null) throw new ArgumentNullException(nameof(comparer));
+            Add(comparer.GetHashCode(value));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        public static int Combine<T1>(T1 value1)
+        {
+            var hc = new HashCode();
+            hc.Add(value1);
```

Comment:
I've once tried combining hash-codes with generic code, but due to performance (factor 2 to 8) ended up manually unrolling all `.Combine()` overloads. I see that you've elegantly improved handling the state and queue, so this might improve the situation; but do you have some ballpark estimate how much this costs? (For comparison: this was a larger difference than choosing SipHash 1-3, which would be hash-dos resistant).

Reply:
I'll benchmark post the results in a comment below.